### PR TITLE
Particle integration in library and example

### PR DESCRIPTION
**Pull request by FelixTUD**
1. Added Particle rendering capability for the library

2. Updated example to use particle rendering and volume rendering

### CODE REVIEW

File: example/example_details.hpp [hunk 1]
```diff
@@ -19,6 +19,28 @@
...
+>
+void update_particles(
+  	THost3 hostBuffer3,
+	TDev3 deviceBuffer3,
+	size_t particle_count,
+	float pos)
+{
+  	for(size_t i = 0; i < particle_count; i++)
+	{
+	  //hostBuffer3[i][0] = 0.0f;
```
Comment (ax3l) [inline example/example_details.hpp]: please remove all dead code, that's what we have `git` for :)

File: example/example.cpp [hunk 2]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+
+#define PARTICLE_COUNT 64
+
+
+
 //////////////////////
 // Example Source 1 //
 //////////////////////
+/*
+ * First volume source
```
Comment (psychocoderHPC) [inline example/example.cpp]: please remove this comment and rename the comment above `Volume Source 1` else both comments will somehow say the same.

File: example/example.cpp [hunk 3]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+        result.value = value;
+        return result;
+    }
 };
 
 //////////////////////
 // Example Source 2 //
 //////////////////////
+/*
+ * Second volume source
```
Comment (psychocoderHPC) [inline example/example.cpp]: same please remove

File: example/example.cpp [hunk 4]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+    }
+
+    ISAAC_HOST_DEVICE_INLINE ElemType getPosition() const
+    {
+        return *current_element;
+    }
+
+    ISAAC_HOST_DEVICE_INLINE isaac_float_dim<feature_dim> getAttribute() const
+    {
+        return {isaac_float ( local_grid_coord.x ), isaac_float ( local_grid_coord.y ), isaac_float ( local_grid_coord.z ) };
```
Comment (psychocoderHPC) [inline example/example.cpp]: please remove the space after `isaac_float` and before `(...)`

File: example/example.cpp [hunk 5]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
-	{
-		srand(time(NULL));
-		id = rand() % 1000000;
-	}
+    //Settings the parameters for the example
+    char __server[] = "localhost";
+    char* server = __server;
+    char* filename = NULL;
+    //If existend first parameter is the server. Default: "localhost"
+    if ( argc > 1 ) {
```
Comment (psychocoderHPC) [inline example/example.cpp]: could you remove `{}` https://github.com/ComputationalRadiationPhysics/contributing/blob/master/codingGuideLines/cpp.md#12-if-statements

File: example/example.cpp [hunk 6]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+        server = argv[1];
+    }
+    int port = 2460;
+    //If existend second parameter is the port. Default: 2460
+    if ( argc > 2 ) {
```
Comment (psychocoderHPC) [inline example/example.cpp]: same `{}`

File: example/example.cpp [hunk 7]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+        port = atoi ( argv[2] );
+    }
+    if ( argc > 3 ) {
```
Comment (psychocoderHPC) [inline example/example.cpp]: same `{}`

File: example/example.cpp [hunk 8]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+#if ISAAC_ALPAKA == 1
+    ////////////////////////////////////
+    // Alpaka specific initialization //
+    ////////////////////////////////////
+    using AccDim = alpaka::dim::DimInt<3>;
+    using SimDim = alpaka::dim::DimInt<3>;
+    using DatDim = alpaka::dim::DimInt<1>;
+
+    //using Acc = alpaka::acc::AccGpuCudaRt<AccDim, ISAAC_IDX_TYPE>;
+    //using Stream  = alpaka::stream::StreamCudaRtSync;
```
Comment (psychocoderHPC) [inline example/example.cpp]: please revert this change. IN alapka dev stream was renamed to `queue`

File: example/example.cpp [hunk 9]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+    using Acc = alpaka::acc::AccCpuOmp2Blocks<AccDim, ISAAC_IDX_TYPE>;
+    using Stream  = alpaka::stream::StreamCpuSync;
```
Comment (psychocoderHPC) [inline example/example.cpp]: please revert

File: example/example.cpp [hunk 10]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+        stream, //Alpaka specific stream instance
+#endif
+        name, //Name of the visualization shown to the client
+        0, //Master rank, which will opens the connection to the server
+        server, //Address of the server
+        port, //Inner port of the server
+        framebuffer_size, //Size of the rendered image
+        global_size, //Size of the whole volumen including all nodes
+        local_size, //Local size of the subvolume
+    {PARTICLE_VOLUME_X,PARTICLE_VOLUME_Y,PARTICLE_VOLUME_Z},
```
Comment (psychocoderHPC) [inline example/example.cpp]: the indention is missing. Could you please also indent all other arguments until `)`

File: example/example.cpp [hunk 11]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+    position, //Position of the subvolume in the globale volume
+    particle_sources,
+    sources, //instances of the sources to render
+    scaling
+    );
+
+    //Setting up the metadata description (only master, but however slaves could then add metadata, too, it would be merged)
+    if ( rank == 0 ) {
```
Comment (psychocoderHPC) [inline example/example.cpp]: Please revert this change `{` must be on an own line https://github.com/ComputationalRadiationPhysics/contributing/blob/master/codingGuideLines/cpp.md#12-if-statements

File: example/example.cpp [hunk 12]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+    int merge_time = 0;
+    int kernel_time = 0;
+    int copy_time = 0;
+    int video_send_time = 0;
+    int buffer_time = 0;
+    bool pause = false;
+    //How often should the visualization be updated?
+    int interval = 1;
+    int step = 0;
+    if ( rank == 0 ) {
```
Comment (psychocoderHPC) [inline example/example.cpp]: please revert `{}`

File: example/example.cpp [hunk 13]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+    ///////////////
+    while ( !force_exit ) {
+        //////////////////
+        // "Simulation" //
+        //////////////////
+        if ( !pause ) {
+            a += 0.01f;
+            int start_simulation = visualization->getTicksUs();
+#if ISAAC_NO_SIMULATION == 0
+            if ( !filename ) {
```
Comment (psychocoderHPC) [inline example/example.cpp]: please revert `{}`

File: example/example.cpp [hunk 14]
```diff
@@ -24,507 +24,627 @@ using namespace isaac;
...
+#endif
+
+    MPI_Finalize();
+    return 0;
 }
 
 // Not necessary, just for the example
 
 
+
```
Comment (psychocoderHPC) [inline example/example.cpp]: please remove all empty lines expect the last

File: lib/ISAACConfig.cmake [hunk 15]
```diff
@@ -219,7 +219,7 @@ FIND_PACKAGE_HANDLE_STANDARD_ARGS(ISAAC
                                         ISAAC_INCLUDE_DIRS
                                         JANSSON_LIBRARIES
                                         JANSSON_INCLUDE_DIRS
-                                        CMAKE_THREAD_LIBS_INIT
+                                        Threads_FOUND
```
Comment (psychocoderHPC) [inline lib/ISAACConfig.cmake]: Could you make a separate PR for this one line change. We need to know if this is a bug and if we need to mark the change as bug to be able to write a changelog.

File: lib/isaac.hpp [hunk 16]
```diff
@@ -68,1792 +69,1948 @@ template <
...
+            typename TParticleSource,
+            typename TWeight,
+            typename TOffset,
+            typename TPointer
+            >
+        ISAAC_HOST_INLINE  void operator() (
+            const int I,
+            TParticleSource& particle_source,
+            const TWeight& weight,
+            const TOffset& weightArrayOffset,
```
Comment (psychocoderHPC) [inline lib/isaac.hpp]: could you please add a doxygen string to the function and give a short description what the arguments mean.

File: lib/isaac.hpp [hunk 17]
```diff
@@ -68,1792 +69,1948 @@ template <
...
+                    alpaka::exec::create<TAcc>
+                    (
+                        workdiv,
+                        kernel,
+                        source,
+                        pointer_array.pointer[ I ],
+                        local_size_array
+                    )
+                );
+                alpaka::stream::enqueue(stream, instance);
```
Comment (psychocoderHPC) [inline lib/isaac.hpp]: please revert and test the changes later on with `alpaka/dev`

File: lib/isaac.hpp [hunk 18]
```diff
@@ -68,1792 +69,1948 @@ template <
...
+                        workdiv,
+                        kernel,
+                        source,
+                        I,
+                        alpaka::mem::view::getPtrNative ( local_minmax ),
+                        local_size_array,
+                        pointer_array.pointer[ I ]
+                    )
+                );
+                alpaka::stream::enqueue(stream, instance);
```
Comment (psychocoderHPC) [inline lib/isaac.hpp]: please revert

File: lib/isaac.hpp [hunk 19]
```diff
@@ -68,1792 +69,1948 @@ template <
...
+        >
+    struct calc_particle_minmax_iterator {
+        template
+        <
+            typename TParticleSource,
+            typename TMinmax,
+            typename TLocalMinmax,
+            typename TLocalSize
+#if ISAAC_ALPAKA == 1
+            ,typename TStream__
```
Comment (psychocoderHPC) [inline lib/isaac.hpp]: Why has the name `__` as postfix?

File: lib/isaac.hpp [hunk 20]
```diff
@@ -68,1792 +69,1948 @@ template <
...
+        auto const instance
+        (
+            alpaka::exec::create<TAcc>
+            (
+                workdiv,
+                kernel,
+                alpaka::mem::view::getPtrNative ( functor_chain_d )
+            )
+        );
+        alpaka::stream::enqueue(stream, instance);
```
Comment (psychocoderHPC) [inline lib/isaac.hpp]: please revert

File: lib/isaac.hpp [hunk 21]
```diff
@@ -68,1792 +69,1948 @@ template <
...
+                isaac_float f_l_height = ( isaac_float ) local_size_scaled[1]/ ( isaac_float ) max_size_scaled * 2.0f;
+                isaac_float f_l_depth = 0.0f;
+                if ( TSimDim::value > 2 ) {
+                    f_l_depth = ( isaac_float ) local_size_scaled[2]/ ( isaac_float ) max_size_scaled * 2.0f;
+                }
+                isaac_float f_x = ( isaac_float ) position_scaled[0]/ ( isaac_float ) max_size_scaled * 2.0f - ( isaac_float ) global_size_scaled[0]/ ( isaac_float ) max_size_scaled;
+                isaac_float f_y = ( isaac_float ) position_scaled[1]/ ( isaac_float ) max_size_scaled * 2.0f - ( isaac_float ) global_size_scaled[1]/ ( isaac_float ) max_size_scaled;
+                isaac_float f_z = 0.0f;
+                if ( TSimDim::value > 2 ) {
+                    f_z = ( isaac_float ) position_scaled[2]/ ( isaac_float ) m
```
Comment (psychocoderHPC) [inline lib/isaac.hpp]: please revert

File: lib/isaac/isaac_defines.hpp [hunk 22]
```diff
@@ -54,7 +54,7 @@
 #endif
 
 #define ISAAC_MAX_RECEIVE 262144 //256kb
-#define ISAAC_Z_NEAR 1.0f
```
Comment (psychocoderHPC) [inline lib/isaac/isaac_defines.hpp]: why do you change this value? Could you give me an explanation of what this variable is doing?

Comment (FelixTUD) [reply lib/isaac/isaac_defines.hpp]: This value defines the distance to the camera frustums near plane. It is usually a small float value and in that specific instance it defines the starting position of the raycasting ray.
For the particle rendering was a value of 1.0 too far away and particles which were near the camera were cut on the near plane.
In the volume rendering were no artifacts, because it isn't visually obvious where the ray starts.

canvas_clipping

File: lib/isaac/isaac_kernel.hpp [hunk 23]
```diff
@@ -1131,28 +2284,36 @@ struct IsaacRenderKernelCaller
                         clipping \
                     ) \
                 ); \
-                alpaka::queue::enqueue(stream, instance); \
+                alpaka::stream::enqueue(stream, instance); \
```
Comment (psychocoderHPC) [inline lib/isaac/isaac_kernel.hpp]: THis is a BUG and must be fixed.
